core/src/Borrower.sol

```diff
@@ -82,11 +95,38 @@ contract Borrower is IUniswapV3MintCallback {
                            MAIN ENTRY POINTS
     //////////////////////////////////////////////////////////////*/
 
+    /**
+     * @notice Warns the borrower that they're about to be liquidated. NOTE: Liquidators are only
+     * forced to call this in cases where the 5% swap bonus is up for grabs.
+     */
+    function warn() external {
+        require(state == State.Ready && unleashLiquidationTime == 0);
+
+        {
+            // Fetch prices from oracle
+            Prices memory prices = getPrices();
+            // Withdraw Uniswap positions while tallying assets
+            Assets memory assets = _getAssets(positions.read(), prices, false);
+            // Fetch liabilities from lenders
+            (uint256 liabilities0, uint256 liabilities1) = _getLiabilities();
+            // Ensure only unhealthy accounts get warned
+            require(!BalanceSheet.isHealthy(prices, assets, liabilities0, liabilities1), "Aloe: healthy");
+        }
+
+        unleashLiquidationTime = uint88(block.timestamp + LIQUIDATION_GRACE_PERIOD);
+
+        emit Warn();
```

Comment:
Nice! What are your thoughts on adding information about amount of debt they need to may back in either token 0 or token 1 to the event? Could that be useful?

Reply:
I was just writing about this as well. I think it would be good to add information about the warning in the event.

Reply:
It'd definitely be useful, but those amounts (exactly how much they need to repay to avoid liquidation) aren't well-defined. We could just emit what `liabilities0` and `liabilities1` are, but that's pretty easy for them to check themselves so I'm not sure it's worth the gas.

Very very long-term, it'd be cool to make the protocol more aware of when and how different borrowers are going to be liquidated, so that it can plan ahead of time to execute swaps with less price impact. I suspect those computations will require some off-chain math with ZK proofs to be feasible though.

Reply:
What about just general information?

Reply:
Discussed offline. For now it's fine as-is, but if we ever do a dynamic liquidation grace period, it'd be good to emit the deadline in this event.